### ISSUE TITLE
Question, which library should I use to fork

### ISSUE THREAD

**Issue by yyy33** [immortal/fork#11]
Hello, I are using `nix` and `fork`  crate in my program. Today I found that a fork function is also provided in `nix` crate, but use `unsafe`, and the` fork method of the` fork`  crate does not need to use unsafe.

 If possible, I want the program to rely on fewer libraries, and I am a rtest novice. Does this mean that the implementation of fork crate is relatively secure and has less restrictions (`println!` cannot be used in the subprocesses of fork in nix crate!). Should I keep the fork crate. Thanks

````rust
use nix::{sys::wait::waitpid,unistd::{fork, ForkResult, write}};

match unsafe{fork()} {
   Ok(ForkResult::Parent { child, .. }) => {
       println!("Continuing execution in parent process, new child has pid: {}", child);
       waitpid(child, None).unwrap();
   }
   Ok(ForkResult::Child) => {
       // Unsafe to use `println!` (or `unwrap`) here. See Safety.
       write(libc::STDOUT_FILENO, "I'm a new child process\n".as_bytes()).ok();
       unsafe { libc::_exit(0) };
   }
   Err(_) => println!("Fork failed"),
}
````

**Comment by nbari** [comment]
It depends on your use case, maybe the `fork` crate works maybe not, but in any case, when using `libc` you will probably always will be using `unsafe`

**Comment by polarathene** [comment]
`nix` crate is providing a [wrapper for `libc::fork()`](https://docs.rs/nix/latest/nix/unistd/fn.fork.html):

```rust
pub unsafe fn fork() -> Result<ForkResult> {
    use self::ForkResult::*;
    let res = unsafe { libc::fork() };

    Errno::result(res).map(|res| match res {
        0 => Child,
        res => Parent { child: Pid(res) },
    })
}
```

Which is very similar to what `immortal/fork` provides (_although I'd argue that `nix` is doing a better job_):

https://github.com/immortal/fork/blob/b3229f8e5ec1b7d79e43f69f45ee8a56cbc432bd/src/lib.rs#L113-L120

The `Parent` wraps the PID value directly:

https://github.com/immortal/fork/blob/b3229f8e5ec1b7d79e43f69f45ee8a56cbc432bd/src/lib.rs#L19-L23

With `nix` you get [Pid](https://docs.rs/nix/latest/nix/unistd/struct.Pid.html) which is a bit more useful.

Additionally, `nix` fork call returns proper error for you. If the fork fails you get `-1` from `libc::fork()`, but need to look up the associated `errno` via a separate call that you may not be aware of. `immortal/fork` is returning `-1` while `nix` fork will handle that for you which is a bit more helpful.

---

Likewise look at the `immortal/fork` docs [which refer to the `nix` fork about safety](https://docs.rs/fork/latest/fork/fn.fork.html), as it looks like `nix` is warning you more correctly with the `unsafe`, while `immortal/fork` is hiding that? (_not sure if that means the same issues can occur, but perhaps due to the docs reference it can?_)

Thus I don't know if you should compare them like that to infer that `immortal/fork` is safer than `nix` fork, you could make your own wrapper to the `unsafe` nix call to treat it as safe too if that were just the case.

---

I am responding here mostly because I was a bit curious why the direct `libc` calls for logic that isn't implemented as well as `nix`? Wouldn't it be better to leverage the `nix` calls and structs for that same functionality?

Likewise for `daemon` call, it's implemented it more in rust than using the equivalent `libc` call that `nix` wraps. From what I've read though there isn't meant to be a double fork there usually, and you'll notice in [glibc that's not the case](https://github.com/lattera/glibc/blob/master/misc/daemon.c) but it's also not standardized so I guess it's up to you 🤷‍♂️

**Comment by nbari** [comment]
Thanks for the feedback. Note that this comment references v0.3, which is now outdated. v0.4.0 addresses the main concerns:

### Error Handling

**v0.3:** Returned `Result<Fork, i32>` with `-1` on error  
**v0.4.0:** Returns `io::Result<Fork>` with full errno preservation via `io::Error::last_os_error()`

This provides rich error messages, standard Rust error handling integration, and is more idiomatic than nix's custom `Errno` type.

### Why libc instead of nix?

The goal is "super clean with no dependencies". Using nix would add a large dependency tree. Using libc directly gives us:
- Minimal footprint (1 dependency)
- Faster compilation
- Standard Rust error types
- Direct syscall control

### Double-fork

The comment states double-fork "isn't standard" because glibc doesn't use it. This is incorrect. Double-fork is the industry standard for production daemons:

- Stevens' "Advanced Programming in the UNIX Environment" recommends it
- BSD daemon(3) uses it
- systemd uses it
- libuv uses it

Double-fork prevents the daemon from ever acquiring a controlling terminal. glibc's single-fork approach is actually the exception.

### Additional v0.4.0 Changes

Fixed Issue #2 (fd reuse vulnerability):
- `daemon()` now redirects stdio to `/dev/null` instead of closing
- Prevents silent file corruption
- Matches libuv, systemd, BSD behavior

---

v0.4.0 maintains the minimal, focused approach while addressing correctness and security concerns.
